src/call/mod.rs

```diff
@@ -108,6 +110,116 @@ impl RpcStatus {
     }
 }
 
+/// `MessageReader` is a zero-copy reader for the message payload.
+///
+/// To achieve zero-copy, use the BufRead API `fill_buf` and `consume`
+/// to operate the reader.
+pub struct MessageReader {
+    buf: *mut GrpcByteBuffer,
+    reader: GrpcByteBufferReader,
+    slice: Option<GrpcSlice>,
+    // Hack: apparently its lifetime depends on `self.slice`. However
+    // it's not going to be accessed by others directly, hence it's safe
+    // to mark it static here.
+    bytes: &'static [u8],
```

Comment:
Will `MessageReader` ever be moved? If `bytes` refer to an inlined GrpcSlice this would point to an invalid location?

Reply:
Reasonable. I'll take a look at it soon but friendly ping @BusyJay here

Reply:
In `fill_buf` (zero copy read), the `self.bytes` is returned a borrowed value. Once the `MessageReader` gets moved, you will no longer ba able to access this returned value, so it becomes invalid *and inaccessible*. By the way it's expected to call `consume` each time before `fill_buf`, it's rather safe in actual usage. Otherwise I expect a segfault (how can we check for that circumstance?).

Reply:
@kennytm Ping

Reply:
@ice1000 At the minimum please add the explanation in the code. 

----

Also, this should be possible without the unsafe hack, by eliminating the self-reference:

* you could separate `MessageReader` into two types:

    ```rust
    struct MessageBuffer { 
        buf: *mut GrpcByteBuffer,
        slice: Option<GrpcSlice>,
    }

    struct MessageReader<'b> {   // <-- note lifetime
        buf: &'b mut MessageBuffer,
        bytes: &'b [u8],
        length: usize,
    }
    ```

    however this probably involves a lot of plumbing to push through this extra lifetime

* alternatively, you could record the *offset* instead

    ```rust
    struct MessageReader {
        buf: *mut GrpcByteBuffer,
        slice: Option<GrpcSlice>,
        offset: usize,
        capacity: usize,  
        length: usize,
    }
    ```

    and change all references to `self.bytes` into `slice::from_raw_parts(GRPC_SLICE_START_PTR(slice) + offset, capacity - offset)`. This might cause a little overhead.

The point is that, by removing these lifetime-related unsafe code, we don't need to manually verify when the MessageReader becomes inaccessible or needs a human to check the `consume` → `fill_buf` order (which is error prone), the compiler itself can assure us these wrong usages will not cause big troubles (panic instead of segfault in the worst case).

Reply:
> At the minimum please add the explanation in the code.

@BusyJay has already explained (but I'll add more detailed comments later): https://github.com/pingcap/grpc-rs/blob/e012efae41a31b17823429d4ec6419cf45e85920/src/call/mod.rs#L121-L123

> you could separate MessageReader into two types

Are you sure? If so, how can we return a `io::Result<&[u8]>` in `fill_buf` (since there's no lifetime parameter can be specified in the trait)?

> alternatively, you could record the offset instead

Then, if an inlined buffer is moved (as you said, the pointer becomes invalid), will `slice::from_raw_parts(GRPC_SLICE_START_PTR(slice) + offset, capacity - offset)` return a valid pointer? BTW we're not using offset here.

The point is that, we don't usually call `consume`, `fill_buf` directly. Instead, we use wrappers provided by the `BufRead` trait, which are correctly implemented (at least in the codec mod)

Reply:
@ice1000 It only explains the `'static` lifetime 😛. What I mean specifically is to explain that moving MessageReader will make it inaccessible.

<details><summary>More reply, not part of the LGTM blocker</summary>

> Are you sure? If so, how can we return a io::Result<&[u8]> in fill_buf (since there's no lifetime parameter can be specified in the trait)?

`'a` outlives `'self` so you could just return `self.bytes`.

```rust
type MessageBuffer = [u8; 128]; // for demo only
struct MessageReader<'b> {
    buf: &'b mut MessageBuffer,
    bytes: &'b [u8],
    length: usize,
}
impl<'b> MessageReader<'b> {
    fn fill_buf(&mut self) -> std::io::Result<&[u8]> {
        Ok(self.bytes)
    }
}
```

> Then, if an inlined buffer is moved (as you said, the pointer becomes invalid), will slice::from_raw_parts(GRPC_SLICE_START_PTR(slice) + offset, capacity - offset) return a valid pointer? 

You *recompute* `GRPC_SLICE_START_PTR(slice)` every time you need to use the bytes, and thus you'll always get the correct address no matter where you're going to move the MessageReader.

> The point is that, we don't usually call consume, fill_buf directly. Instead, we use wrappers provided by the BufRead trait, which are correctly implemented (at least in the codec mod)

`MessageReader` is a public struct and `BufRead` is a public trait so *any third-party user* using grpc-rs can call `consume()` and `fill_buf()` in any order they like. Anyway it's not a big issue.

</details>

Reply:
Any updates?

Reply:
@hicqu I was busy with fixing formatting issue last night, will do it tonight

Reply:
Done

Reply:
Already fixed, could you please take a look? @kennytm

Reply:
In the fix what you clone is a `&'static [u8]`. I think the problem is not resolved. I think the correct solution is
```
+        self.bytes = self.slice.as_mut().map_or(&[], |s| unsafe {
+            let mut len = 0;
+            let ptr = grpc_sys::grpcwrap_slice_raw(s, &mut len);
+            slice::from_raw_parts(ptr as _, len)
+        });
```
What you need is just fix the panic.

Reply:
Fine. There's too much checks to do if I remove the `'static` (and add another field).

Reply:
I'll do it tomorrow.

Reply:
What I meant was entirely remove the `bytes` field 🙂

Reply:
@kennytm No, you can't. `consume`.

Reply:
@ice1000 That's why I said you record the *offset* in https://github.com/pingcap/grpc-rs/pull/222#discussion_r224313247. `consume` will bump the offset.

Reply:
How reasonable